### ISSUE TITLE
iOS crashes after too many errors

### ISSUE THREAD

**Issue by metalaureate** [terikon/cordova-plugin-photo-library#13]
I'm trying to get this library to work for the first time on Ionic.

I run the sample code inside my Ionic app built on an iOS 10 device, connected to Xcode:

```
cordova.plugins.photoLibrary.getLibrary(
        function (library) {
          $log.debug('library',library);
          // Here we have the library as array

          library.forEach(function(libraryItem) {
            $log.debug(libraryItem.id);          // Id of the photo
            $log.debug(libraryItem.photoURL);    // Cross-platform access to photo
            $log.debug(libraryItem.thumbnailURL);// Cross-platform access to thumbnail
            $log.debug(libraryItem.filename);
            $log.debug(libraryItem.width);
            $log.debug(libraryItem.height);
            $log.debug(libraryItem.creationDate);
            $log.debug(libraryItem.nativeURL);   // Do not use, as it is not accessible on all platforms;

          });

        },
        function (err) {
          $log.error('Error occured',err);
        },
        {
          thumbnailWidth: 512,
          thumbnailHeight: 384,
          quality: 0.8
        }
      );
```

In the Xcode console, the following errors appear, and after about 30 secs Xcode 'loses connection' with the app because it quits. I guess this could be iOS killing the app for generating too many errors? But how do I avoid these errors (many hundreds are generated until the app crashes):

> Failed to load image data for asset <PHAsset: 0x100c896c0> 7BE3559A-45F5-4182-8635-C81284D4CF03/L0/001 mediaType=1/0, sourceType=1, (2448x3264), creationDate=2016-07-13 23:09:13 +0000, location=0, hidden=0, favorite=0  with format 9999
> 2016-11-12 11:57:23.065292 Hisgeogram[6153:1488288] [Generic] Failed to load image data for asset <PHAsset: 0x100c89d50> 76494272-6B94-4053-91E9-D08C78F0F5D3/L0/001 mediaType=1/0, sourceType=1, (2448x3264), creationDate=2016-07-14 02:23:27 +0000, location=0, hidden=0, favorite=0  with format 9999

I have no idea what is going on, appreciate any tips.

Cordova CLI: 6.4.0
Gulp version:  CLI version 3.9.1
Gulp local:  
Ionic Framework Version: 1.3.1
Ionic CLI Version: 2.0.0
Ionic App Lib Version: 2.0.0
ios-deploy version: 1.8.6 
ios-sim version: 5.0.8 
OS: Mac OS X El Capitan
Node Version: v6.9.1
Xcode version: Xcode 8.1 Build version 8B62

**Comment by viskin** [comment]
According to [this](http://stackoverflow.com/questions/28170444/photos-framework-requestimagedataforasset-occasionally-fails), these error might happen when fetching images in an iCloud photo stream (Shared Album) that has not been downloaded, and options.synchronous = YES.
Proposed solution is using synchronous = NO. Or just exclude Shared Album at all.
Can you confirm that you have photos in Shared Album?

**Comment by viskin** [comment]
Can you please also modify [following line](https://github.com/terikon/cordova-plugin-photo-library/blob/master/src/ios/PhotoLibraryService.swift#L39) in PhotoLibraryService.swift, and remove .typeCloudShared from it:

      fetchOptions.includeAssetSourceTypes = [.typeUserLibrary, .typeiTunesSynced, .typeCloudShared]

If it will solve the crash, we'll know that Shared Album is exactly the problem.

**Comment by metalaureate** [comment]
Thanks for helping me out! I removed both `.typeiTunesSynced` and `.typeCloudShared`for good measure. The same problem occurs:


> Hisgeogram[7361:1672099] [Generic] Failed to load image data for asset <PHAsset: 0x100cc1d90> AC81F08E-47C5-4F1D-8E7C-9BAD4DCD195A/L0/001 mediaType=1/0, sourceType=1, (2448x3264), creationDate=2016-06-12 19:15:51 +0000, location=1, hidden=0, favorite=0  with format 9999
> 2016-11-13 08:27:35.049690 Hisgeogram[7361:1672303] [Generic] Failed to load image data for asset <PHAsset: 0x100cc7040> 4081EBAD-E825-4F7E-928E-F82806CF888B/L0/001 mediaType=1/0, sourceType=1, (2448x3264), creationDate=2016-06-15 01:51:23 +0000, location=1, hidden=0, favorite=0  with format 9999
> 2016-11-13 08:27:35.051610 Hisgeogram[7361:1672316] [Generic] Failed to load image data for asset <PHAsset: 0x100cc4b80> A402879F-99E7-4445-851D-04C10BE54CBB/L0/001 mediaType=1/0, sourceType=1, (2448x3264), creationDate=2016-06-15 02:32:12 +0000, location=1, hidden=0, favorite=0  with format 9999

**Comment by viskin** [comment]
I'll try to run with synchronous = NO, as advised by article above. And I'll try to scan Internet for similar errors.

**Comment by metalaureate** [comment]
I'm a dummy but let me know if I can help.

**Comment by viskin** [comment]
I converted getLibrary code to be asynchronous.
Modified code is in [getLibrary-async](https://github.com/terikon/cordova-plugin-photo-library/tree/getLibrary-async) branch.
Please run it on your device. If it resolves the problem, I will merge the code to main branch.

**Comment by metalaureate** [comment]
Build failed. After running

```
cordova plugin rm cordova-plugin-photo-library 
cordova plugins add https://github.com/terikon/cordova-plugin-photo-library.git#bd87c62   --save
ionic build ios
```

> The following build commands failed:
>         CompileSwift normal x86_64 /Users/shill/Projects/hisgeogram/hisgeogram-app/hisgeogram/platforms/ios/Hisgeogram/Plugins/cordova-plugin-photo-library/PhotoLibrary.swift
> 
>         CompileSwiftSources normal x86_64 com.apple.xcode.tools.swift.compiler
> (2 failures)
> 
> Error: Error code 65 for command: xcodebuild with args: -xcconfig,/Users/shill/Projects/hisgeogram/hisgeogram-app/hisgeogram/platforms/ios/cordova/build-debug.xcconfig,-workspace,Hisgeogram.xcworkspace,-scheme,Hisgeogram,-configuration,Debug,-sdk,iphonesimulator,-destination,platform=iOS Simulator,name=iPhone 5s,build,CONFIGURATION_BUILD_DIR=/Users/shill/Projects/hisgeogram/hisgeogram-app/hisgeogram/platforms/ios/build/emulator,SHARED_PRECOMPS_DIR=/Users/shill/Projects/hisgeogram/hisgeogram-app/hisgeogram/platforms/ios/build/sharedpch
>

**Comment by viskin** [comment]
Sorry my bad. Forgot to check-in one file. Please re-try with df27a7af9f68f89d81201633755bcb4a7df9d279

**Comment by metalaureate** [comment]
This worked! Thank you. It took 10 minutes to retrieve 2,000 images, which is fairly low volume for an average consumer phone. Is there any way to retrieve a subset, or paginate, or trigger a progress event?

**Comment by viskin** [comment]
Good, I will merge it.
10 minutes for call to getLibrary? I guess you have all these photos in cloud. But in such case it shouldn't download photos over network. I'd prefer to just not show them. Or add progress event, as you say.
What happened when you open standard Photos app? Did these 2000 photos appeared there?

**Comment by metalaureate** [comment]
Yes. I believe they are all on my phone. I will modify the filter criteria to remove cloud photos and report back later.

**Comment by viskin** [comment]
Meanwhile, I opened #14 to track the issue. It seems necessary to provide partial results if retrieving the whole library takes more than a second.

**Comment by metalaureate** [comment]
Removing iCloud and iTunes sync options reduced the response time to about 10 secs and mysteriously retrieved the same 2,096 results.

**Comment by viskin** [comment]
Maybe following lines in [PhotoLibraryService.swift](https://github.com/terikon/cordova-plugin-photo-library/blob/master/src/ios/PhotoLibraryService.swift) cause so slow access:

thumbnailRequestOptions.isNetworkAccessAllowed = true
imageRequestOptions.isNetworkAccessAllowed = true

Can you try set **false** for both of them on your device, while iCloud and iSync options are both on?

**Comment by metalaureate** [comment]
Yes that works. 7 secs.

**Comment by viskin** [comment]
Ok, so I will set isNetworkAccessAllowed to false. 7 sec is still alot, what do you think about following improvement:
getLibrary will get additional parameter onPartialResult:
```      
cordova.plugins.photoLibrary.getLibrary(function onComplete (library) {},  function onError (err) {}, options, function onPartialResult(partialLibrary) {});
```

onPartialResult will be called by timer each say 0.5 sec, until photoLibrary completes and onComplete is  called. onPartialResult will receive partialLibrary parameter which is array of library items that are retrieved by then. partialLibrary will grow each 0.5 sec.

**Comment by metalaureate** [comment]
I think that would work great.

**Comment by viskin** [comment]
Ok, I will try to implement this in very near future.

**Comment by metalaureate** [comment]
Greatly appreciated, thank you.

**Comment by viskin** [comment]
@metalaureate can you please try this new partial result feature?
It is in https://github.com/terikon/cordova-plugin-photo-library.git#getLibrary-PartialResult
It will provide partial results each half a second.

**Comment by viskin** [comment]
Merged to 1.2.0

**Comment by santhoshs5** [comment]
I am not using your library but occasionally some times it is throwing the same error

can you guys tell me the fix.
